src/api.py

```diff
@@ -1,36 +1,45 @@
+import json
 import logging
 import os
 import signal
+from logging.handlers import TimedRotatingFileHandler
 from queue import Queue
+from threading import Thread, Condition
 
 import tornado
 import tornado.ioloop
 import tornado.web
-from bson import ObjectId
+from tornado import httpclient
 
+import src.constants.api_keys as api_key
 import src.constants.constants as consts
 import src.constants.db_keys as db_key
-import src.handlers as handlers
 from src.auth import initialize_token
-from src.config import PORT, HEARTBEAT_INTERVAL
-from src.config import GRADER_REGISTER_ENDPOINT, GRADING_JOB_ENDPOINT, GRADING_RUN_ENDPOINT, HEARTBEAT_ENDPOINT
+from src.config import WORKER_REGISTER_ENDPOINT, GRADING_JOB_ENDPOINT, GRADING_RUN_ENDPOINT, HEARTBEAT_ENDPOINT
+from src.config import PORT, HEARTBEAT_INTERVAL, LOGS_DIR
 from src.database import DatabaseResolver
-from src.utilities import get_time, get_string_from_time
+from src.handlers.client_handler import AddGradingRunHandler, GradingRunHandler
+from src.handlers.worker_handler import WorkerRegisterHandler, GradingJobHandler, HeartBeatHandler
+from src.utilities import get_time
 
 # setting up logger
-os.makedirs(consts.LOGS_DIR, exist_ok=True)
+os.makedirs(LOGS_DIR, exist_ok=True)
 logging.basicConfig(
     handlers=[
-        logging.FileHandler('{}/{}.log'.format(consts.LOGS_DIR, get_string_from_time())),
+        TimedRotatingFileHandler('{}/log'.format(LOGS_DIR), when='midnight', backupCount=7),
         logging.StreamHandler()
     ],
     level=logging.INFO
 )
 logger = logging.getLogger()
+heartbeat_running = True
+heartbeat_cv = Condition()
```

Comment:
Could we have these be objects bound to the `shutdown` and the `heartbeat_validator` calls? Logging is generally okay as a global variable but once we have application logic variables as globals, we may want to flip those to be bound.

Reply:
Should be resolved now. Created a Periodic Callback Thread object which actually helped clean things up better and more safely.